### ISSUE TITLE
0.73 crashing most of the time

### ISSUE THREAD

**Issue by alphazo** [sshuttle/sshuttle#50]
I thought I was only impacted by bug https://github.com/sshuttle/sshuttle/issues/31 but I'm now seeing a constant crash when using 0.73.20151209-1 (ArchLinux) while 0.72 works flawlessly.

Here is the log for 0.72. The tunnel opens but when I go somewhere, like ipleak.net it instantly crashes:
```
# sshuttle --dns -v -r mybox 0/0
Starting sshuttle proxy.
firewall manager: Starting firewall with Python version 3.5.0
firewall manager: ready method name nat.
UDP enabled: False
TCP redirector listening on ('127.0.0.1', 12300).
DNS listening on ('127.0.0.1', 12300).
Starting client with Python version 3.5.0
c : connecting to server...
Starting server with Python version 2.7.3
 s: latency control setting = True
 s: available routes:
 s:   2/192.168.1.100/32
 s:   2/10.8.0.2/32
 s:   2/192.168.1.104/32
 s:   2/62.210.75.0/24
 s:   2/10.8.0.0/24
c : Connected.
firewall manager: setting up.
>> iptables -t nat -N sshuttle-12300
>> iptables -t nat -F sshuttle-12300
>> iptables -t nat -I OUTPUT 1 -j sshuttle-12300
>> iptables -t nat -I PREROUTING 1 -j sshuttle-12300
>> iptables -t nat -A sshuttle-12300 -j RETURN --dest 127.0.0.0/8 -p tcp
>> iptables -t nat -A sshuttle-12300 -j REDIRECT --dest 0.0.0.0/0 -p tcp --to-ports 12300 -m ttl ! --ttl 42
>> iptables -t nat -A sshuttle-12300 -j REDIRECT --dest 192.168.0.254/32 -p udp --dport 53 --to-ports 12300 -m ttl ! --ttl 42
c : DNS request from ('192.168.0.204', 53455) to None: 33 bytes
c : DNS request from ('192.168.0.204', 6296) to None: 43 bytes
c : DNS request from ('192.168.0.204', 65112) to None: 32 bytes
c : DNS request from ('192.168.0.204', 27410) to None: 69 bytes
c : DNS request from ('192.168.0.204', 37825) to None: 69 bytes
c : DNS request from ('192.168.0.204', 55368) to None: 69 bytes
c : DNS request from ('192.168.0.204', 6388) to None: 69 bytes
c : DNS request from ('192.168.0.204', 9549) to None: 69 bytes
c : DNS request from ('192.168.0.204', 26288) to None: 69 bytes
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "assembler.py", line 34, in <module>
  File "sshuttle.server", line 338, in main
KeyError: 6
c : DNS request from ('192.168.0.204', 50164) to None: 69 bytes
c : DNS request from ('192.168.0.204', 2029) to None: 69 bytes
c : DNS request from ('192.168.0.204', 50038) to None: 69 bytes
c : DNS request from ('192.168.0.204', 27739) to None: 69 bytes
c : DNS request from ('192.168.0.204', 56420) to None: 69 bytes
c : DNS request from ('192.168.0.204', 49509) to None: 69 bytes
firewall manager: undoing changes.
>> iptables -t nat -D OUTPUT -j sshuttle-12300
>> iptables -t nat -D PREROUTING -j sshuttle-12300
>> iptables -t nat -F sshuttle-12300
>> iptables -t nat -X sshuttle-12300
c : fatal: other end: ConnectionResetError(104, 'Connection reset by peer')
```
 
Now if install 0.72 and go over the exact same steps it simply works:

```
# sshuttle --dns -v -r mybox 0/0
Starting sshuttle proxy.
UDP support requires tproxy; disabling UDP.
TCP redirector listening on ('127.0.0.1', 12300).
DNS listening on ('127.0.0.1', 12300).
firewall manager ready method nat.
c : connecting to server...
 s: latency control setting = True
 s: available routes:
 s:   2/192.168.1.100/32
 s:   2/10.8.0.2/32
 s:   2/192.168.1.104/32
 s:   2/62.210.75.0/24
 s:   2/10.8.0.0/24
c : connected.
Connected.
firewall manager: starting transproxy.
>> iptables -t nat -N sshuttle-12300
>> iptables -t nat -F sshuttle-12300
>> iptables -t nat -I OUTPUT 1 -j sshuttle-12300
>> iptables -t nat -I PREROUTING 1 -j sshuttle-12300
>> iptables -t nat -A sshuttle-12300 -j RETURN --dest 127.0.0.0/8 -p tcp
>> iptables -t nat -A sshuttle-12300 -j REDIRECT --dest 0.0.0.0/0 -p tcp --to-ports 12300 -m ttl ! --ttl 42
>> iptables -t nat -A sshuttle-12300 -j REDIRECT --dest 192.168.0.254/32 -p udp --dport 53 --to-ports 12300 -m ttl ! --ttl 42
c : DNS request from ('192.168.0.204', 2671) to None: 33 bytes
c : DNS request from ('192.168.0.204', 17970) to None: 43 bytes
c : DNS request from ('192.168.0.204', 53672) to None: 32 bytes
c : DNS request from ('192.168.0.204', 24721) to None: 69 bytes
c : DNS request from ('192.168.0.204', 48649) to None: 69 bytes
c : DNS request from ('192.168.0.204', 29105) to None: 69 bytes
c : DNS request from ('192.168.0.204', 55131) to None: 69 bytes
c : DNS request from ('192.168.0.204', 19418) to None: 69 bytes
c : DNS request from ('192.168.0.204', 32944) to None: 69 bytes
c : DNS request from ('192.168.0.204', 46585) to None: 69 bytes
c : DNS request from ('192.168.0.204', 64757) to None: 69 bytes
c : DNS request from ('192.168.0.204', 29299) to None: 69 bytes
c : DNS request from ('192.168.0.204', 36575) to None: 69 bytes
c : DNS request from ('192.168.0.204', 54017) to None: 69 bytes
c : DNS request from ('192.168.0.204', 18932) to None: 69 bytes
c : DNS request from ('192.168.0.204', 22942) to None: 69 bytes
c : DNS request from ('192.168.0.204', 57166) to None: 69 bytes
c : DNS request from ('192.168.0.204', 4303) to None: 69 bytes
c : DNS request from ('192.168.0.204', 40961) to None: 69 bytes
c : DNS request from ('192.168.0.204', 31485) to None: 69 bytes
c : DNS request from ('192.168.0.204', 33141) to None: 69 bytes
c : DNS request from ('192.168.0.204', 8547) to None: 69 bytes
c : DNS request from ('192.168.0.204', 56833) to None: 69 bytes
c : DNS request from ('192.168.0.204', 65380) to None: 69 bytes
c : DNS request from ('192.168.0.204', 34231) to None: 69 bytes
c : DNS request from ('192.168.0.204', 54681) to None: 69 bytes
c : DNS request from ('192.168.0.204', 41931) to None: 69 bytes
c : DNS request from ('192.168.0.204', 52028) to None: 69 bytes
c : DNS request from ('192.168.0.204', 24173) to None: 69 bytes
c : DNS request from ('192.168.0.204', 60337) to None: 69 bytes
c : DNS request from ('192.168.0.204', 15730) to None: 69 bytes
c : DNS request from ('192.168.0.204', 45951) to None: 69 bytes
c : DNS request from ('192.168.0.204', 39589) to None: 69 bytes
c : DNS request from ('192.168.0.204', 64496) to None: 37 bytes
c : DNS request from ('192.168.0.204', 56684) to None: 39 bytes
c : Accept TCP: 192.168.0.204:41826 -> 192.168.0.254:53.
```

**Comment by brianmay** [comment]
I think this might already have been fixed in git by pull request #46.

**Comment by alphazo** [comment]
Thanks for the heads-up. I confirm that the latest git version v0.73+51_ge433c59-1 seems to fix my problem. sshuttle has been my VPN of choice for many years. I found it to work much better over OpenVPN especially when the connection is not stable.
